File: implementations/java/provekit-claim-envelope/src/main/java/com/provekit/claimenvelope/Cbor.java
// SPDX-License-Identifier: Apache-2.0
//
// Deterministic CBOR encoder. RFC 8949 §4.2.1 rules:
//   - shortest-form integer encoding (smallest of short / u8 / u16 / u32 / u64)
//   - definite-length items only
//   - map keys sorted in bytewise lex order of their CBOR-encoded form
//
// Only emits the major types needed by the proof-envelope shape:
// unsigned int, byte string, text string, array, map.
//
// Mirrors implementations/rust/provekit-proof-envelope/src/cbor.rs and
// implementations/csharp/Provekit.ProofEnvelope/Cbor.cs 1:1.

package com.provekit.claimenvelope;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;

public final class Cbor {
    public static final int MAJOR_UNSIGNED_INT = 0;
    public static final int MAJOR_BYTE_STRING = 2;
    public static final int MAJOR_TEXT_STRING = 3;
    public static final int MAJOR_ARRAY = 4;
    public static final int MAJOR_MAP = 5;

    private Cbor() {}

    /** Append a CBOR head: major-type tag + shortest-form length / value. */
    public static void appendHead(ByteArrayOutputStream out, int major, long arg) {
        if (arg < 0) {
            // We never emit negative integers from this kit; fail loudly.
            throw new IllegalArgumentException("Cbor.appendHead: arg must be non-negative");
        }
        int mt = (major & 0x07) << 5;
        if (arg < 24L) {
            out.write(mt | (int) arg);
            return;
        }
        if (arg <= 0xFFL) {
            out.write(mt | 24);
            out.write((int) arg);
            return;
        }
        if (arg <= 0xFFFFL) {
            out.write(mt | 25);
            out.write((int) ((arg >>> 8) & 0xFF));
            out.write((int) (arg & 0xFF));
            return;
        }
        if (arg <= 0xFFFFFFFFL) {
            out.write(mt | 26);
            out.write((int) ((arg >>> 24) & 0xFF));
            out.write((int) ((arg >>> 16) & 0xFF));
            out.write((int) ((arg >>> 8) & 0xFF));
            out.write((int) (arg & 0xFF));
            return;
        }
        out.write(mt | 27);
        for (int i = 7; i >= 0; i--) {
            out.write((int) ((arg >>> (i * 8)) & 0xFF));
        }
    }

    public static void encodeUint(ByteArrayOutputStream out, long value) {
        appendHead(out, MAJOR_UNSIGNED_INT, value);
    }

    public static void encodeBstr(ByteArrayOutputStream out, byte[] bytes) {
        appendHead(out, MAJOR_BYTE_STRING, bytes.length);
        out.write(bytes, 0, bytes.length);
    }

    public static void encodeTstr(ByteArrayOutputStream out, String utf8) {
        byte[] b = utf8.getBytes(StandardCharsets.UTF_8);
        appendHead(out, MAJOR_TEXT_STRING, b.length);
        out.write(b, 0, b.length);
    }

    public static void encodeArrayHead(ByteArrayOutputStream out, long count) {
        appendHead(out, MAJOR_ARRAY, count);
    }

    public static void encodeMapHead(ByteArrayOutputStream out, long count) {
        appendHead(out, MAJOR_MAP, count);
    }
}

File: implementations/ruby/lib/provekit/cbor.rb
# SPDX-License-Identifier: Apache-2.0
#
# Deterministic CBOR encoder. RFC 8949 §4.2.1 rules:
#   - shortest-form integer encoding (smallest of short / u8 / u16 / u32 / u64)
#   - definite-length items only
#   - map keys sorted in bytewise lex order of their CBOR-encoded form
#   - we emit only the major types we need: unsigned int, byte string,
#     text string, array, map.
#
# Mirrors implementations/rust/provekit-proof-envelope/src/cbor.rs 1:1.

module Provekit
  module Cbor
    # CBOR major type values (RFC 8949 §3.1).
    UNSIGNED_INT = 0
    BYTE_STRING  = 2
    TEXT_STRING  = 3
    ARRAY        = 4
    MAP          = 5

    # Append the CBOR head (major type + argument) to +out+.
    # Selects the shortest-form encoding per RFC 8949 §4.2.1.
    def self.append_head(out, major, arg)
      mt = major << 5
      if arg < 24
        out << (mt | arg)
        return
      end
      if arg <= 0xFF
        out << (mt | 24)
        out << arg
        return
      end
      if arg <= 0xFFFF
        out << (mt | 25)
        out << ((arg >> 8) & 0xFF)
        out << (arg & 0xFF)
        return
      end
      if arg <= 0xFFFF_FFFF
        out << (mt | 26)
        out << ((arg >> 24) & 0xFF)
        out << ((arg >> 16) & 0xFF)
        out << ((arg >> 8) & 0xFF)
        out << (arg & 0xFF)
        return
      end
      out << (mt | 27)
      7.downto(0) { |i| out << ((arg >> (i * 8)) & 0xFF) }
    end

    def self.encode_uint(out, value)
      append_head(out, UNSIGNED_INT, value)
    end

    def self.encode_bstr(out, bytes)
      raw = bytes.is_a?(String) ? bytes.b : bytes.to_s.b
      append_head(out, BYTE_STRING, raw.bytesize)
      raw.each_byte { |b| out << b }
    end

    def self.encode_tstr(out, str)
      raw = str.to_s.dup.force_encoding(Encoding::UTF_8)
      bytes = raw.bytes
      append_head(out, TEXT_STRING, bytes.length)
      bytes.each { |b| out << b }
    end

    def self.encode_array_head(out, count)
      append_head(out, ARRAY, count)
    end

    def self.encode_map_head(out, count)
      append_head(out, MAP, count)
    end

    # Encode a key string for use as a map key.
    # Returns an Array<Integer> of bytes.
    def self.encode_key(key)
      buf = []
      encode_tstr(buf, key)
      buf
    end

    # Pack an array of byte integers into a binary String.
    def self.pack_bytes(byte_array)
      byte_array.pack("C*")
    end
  end
end
